### PR TITLE
[major] Change Button and FocusZone to rootProps model

### PR DESCRIPTION
**Pull request by cliffkoh**
From @dzearing:

> The root problem is this:
>  
> `<div { …this.props }>`
>  
> We want to support scenarios where the caller can provide any default HTML attributes for the root element. A use case might be that they want to provide a data-automation-id or a style attribute. Things like Button or FocusZone are just simple wrappers around simple components and there are many default attributes that could be pushed into the root. But clearly, this is not great if properties that default `div`s don’t support are getting things like “buttonType” and React is warning on it.
>  
> Cliff and I worked out a separate, better solution that will be documented a lot more clearly; for props that you wish to “mix in” to the root, we will expose those as a separate “rootProps” property.
>  
> Something like instead of:
>  
> ```typescript
> <Button style={ { background: ‘red’ } } />
>  ```
>  
> You would say:
>  
> ```typescript
>  <Button rootProps={ { style: ‘background: red’ } } />
>   ```
>  
> While it’s a bit less ergonomic, it means that it is entirely more clear about what things get mixed into the `div`. the `rootProps` property will be of type `HTMLProps<…>` and will have its own description, so that we can be very explicit to say “you can mix any valid div property in here”. That enables the consumer to do things like mix in data-automation-id attributes in a predictable way.
>  
> Every component would get this, we could be really consistent here. We could also support mixing in props into sub elements. For example, perhaps `<TextField>` should have props that get mixed into the input element. You could have `inputProps` and document it appropriately, allowing for more optional flexibility.
>  
> This would definitely be a breaking change but I feel like it’s a flexible but more explicit approach towards extensibility.
> 

Also fixes issue #44.

### CODE REVIEW

**Comment by dzearing** [comment]
@kmees interested in your thoughts here. I think this is probably the correct pattern for allowing the caller to mix in values on key elements within the component. We are currently doing something like <div { ...this.props }> which brings in everything, including unsupported attributes, which React complains about. With this pattern we can namespace off the specific overrides and keep them type-safe.

File: src/components/Button/Button.tsx [hunk 1]
```diff
@@ -30,18 +30,17 @@ export class Button extends React.Component<IButtonProps, IButtonState> {
   }
 
   public render(): JSX.Element {
-    let { buttonType, children, icon, description, ariaLabel, ariaDescription, elementType, disabled } = this.props;
+    let { buttonType, children, icon, description, ariaLabel, ariaDescription, elementType} = this.props;
```
Comment (dzearing) [inline src/components/Button/Button.tsx]: nit space.

File: src/components/Button/Button.tsx [hunk 2]
```diff
@@ -30,18 +30,17 @@ export class Button extends React.Component<IButtonProps, IButtonState> {
...
+    const tag = elementType === ElementType.button ? 'button' : 'a';
 
-    const className = css(this.props.className, 'ms-Button', {
+    const className = css((this.props.rootProps && this.props.rootProps.className), 'ms-Button', {
       'ms-Button--primary': buttonType === ButtonType.primary,
       'ms-Button--hero': buttonType === ButtonType.hero,
       'ms-Button--compound': buttonType === ButtonType.compound,
       'ms-Button--command': buttonType === ButtonType.command,
-      'ms-Button--icon': buttonType === ButtonType.icon,
-      'disabled': (elementType === ElementType.anchor && disabled)
```
Comment (dzearing) [inline src/components/Button/Button.tsx]: It seems like you're removing a disabled css class here... not sure if intended.

File: src/components/CommandBar/CommandBar.tsx [hunk 3]
```diff
@@ -101,7 +101,7 @@ export class CommandBar extends React.Component<ICommandBarProps, ICommandBarSta
     return (
       <div className={ css('ms-CommandBar', className) } ref='commandBarRegion'>
         { searchBox }
-        <FocusZone direction={ FocusZoneDirection.horizontal } role='menubar'>
+        <FocusZone direction={ FocusZoneDirection.horizontal } rootProps={ { role:'menubar' } }>
```
Comment (dzearing) [inline src/components/CommandBar/CommandBar.tsx]: nit space after role:

File: src/demo/components/ExampleCard/ExampleCard.tsx [hunk 4]
```diff
@@ -42,9 +42,9 @@ export class ExampleCard extends React.Component<IExampleCardProps, IExampleCard
         <div className='ExampleCard-header'>
           <span className='ExampleCard-title ms-font-l'>{ title }</span>
           <span className='ExampleCard-toggleButtons ms-font-l'>
-            <Button buttonType={ ButtonType.primary } onClick={ this._onShowExampleClick }>{ isExampleShown ? 'Hide example' : 'Show example' }</Button>
+            <Button buttonType={ ButtonType.primary } rootProps={ { onClick: this._onShowExampleClick } }>{ isExampleShown ? 'Hide example' : 'Show example' }</Button>
```
Comment (dzearing) [inline src/demo/components/ExampleCard/ExampleCard.tsx]: When i read this, i feel confused. Specifically for button. Button does not have a label or wrapper div. It's just a button. So rootProps is ambiguous. It makes me think there is a higher level wrapper around the button.

I expect this to work:

```typescript
<Button onClick={ stuff }>Hi</Button>
```

I think we should explicitly support the common props on the component props, and use the mixin "fooProps" values only for supporting exceptional uncommon mixin cases.

Here's another example:

```typescript
<Checkbox className='foo' label='' checked={ true } onChange={ this._onCheckboxChange } />
```
With the proposed pattern might look like this (keep in mind, the onChange we provide is an extension of the input box's onChange):

```typescript
<Checkbox label='' onChange={ this._onCheckboxChange } inputProps={ { className: 'foo', checked: true } }
```
This feels super weird! Doesn't it?

I believe we should consider using this pattern for exceptional 'mixin' scenarios that are uncommon. For the common things, we pass them through. So:

```typescript
<Checkbox 
  className='foo'
  checked={ true } 
  onChange={ this._onCheckboxChange } 
  inputProps={ { style: { background: 'red' } } } />
```

**Comment by kmees** [comment]
@dzearing I'm a big fan of spreading all additional props (props that are not defined as prop types) to the root component

Simple example from my Overlay component
```jsx
const Overlay = ({ dark, hidden, ...props }) => (
  <div data-fabric="Overlay"
    {...props}
    styleName={cx('ms-Overlay', {
      'ms-Overlay--dark': dark,
      'ms-Overlay--none': hidden
    })} />
)
Overlay.displayName = 'Overlay'
Overlay.propTypes = {
  dark: React.PropTypes.bool,
  hidden: React.PropTypes.bool
}
```

The problem is that typescript doesn't support the [ES Rest/Spread properties](https://github.com/Microsoft/TypeScript/issues/2103) yet. There are some workarounds though (see bottom of issue).

For the components with an `<input />` element, there are basically two ways. Either spread the remaining props into the `<input />` element instead of the root or use `inputProps` as proposed above. Using `inputProps` is probably better for consistency, however there are some 'edge' props like `disabled`, `onFocus` and `onBlur`. Should they be defined on the `<TextField />` (and added to the propTypes) or as `inputProps`? Or maybe support both? Not sure about that.

Example
```jsx
const MyInput = ({ value, onChange, disabled, className = '', inputProps, ...props }) => (
  <div {...props} className={`my-input ${className}`}>
    <input {...inputProps}
      value={value}
      onChange={onChange}
      disabled={disabled || inputProps.disabled} />
  </div>
)
```

File: src/components/Button/Button.tsx [hunk 5]
```diff
@@ -62,13 +62,14 @@ export class Button extends React.Component<IButtonProps, IButtonState> {
...
       tag,
-      assign({
-        'aria-label': ariaLabel,
-        'aria-labelledby': ariaLabel ? null : labelId,
-        'aria-describedby': ariaDescription ? ariaDescriptionId : description ? descriptionId : null,
-        'ref': (c: HTMLButtonElement): HTMLButtonElement => this._buttonElement = c
-      },
-        this.props,
+      assign(
+        this.props.rootProps,
```
Comment (dzearing) [inline src/components/Button/Button.tsx]: I think you should add a {} to the beginning here, otherwise you're mixing things into the rootProps that the user passes in, which could be empty.

File: src/components/ContextualMenu/ContextualMenu.tsx [hunk 6]
```diff
@@ -188,11 +188,13 @@ export class ContextualMenu extends React.Component<IContextualMenuProps, IConte
         <div ref={ (host: HTMLDivElement) => this._host = host} id={ id } className={ css('ms-ContextualMenu-container', className) }>
           { (items && items.length) ? (
             <FocusZone
-              className={ 'ms-ContextualMenu is-open'}
               direction={ FocusZoneDirection.vertical }
-              role='menu'
               ariaLabelledBy={ labelElementId }
               ref={ (focusZone) => this._focusZone = focusZone }
+              rootProps= { {
```
Comment (dzearing) [inline src/components/ContextualMenu/ContextualMenu.tsx]: nit space

File: src/components/FocusZone/FocusZone.tsx [hunk 7]
```diff
@@ -85,12 +85,12 @@ export class FocusZone extends React.Component<IFocusZoneProps, {}> {
...
   public render() {
-    let { className, ariaLabelledBy } = this.props;
+    let { rootProps, ariaLabelledBy } = this.props;
 
     return (
       <div
-        { ...this.props as any }
-        className={ css('ms-FocusZone', className) }
+        { ...this.props.rootProps as any }
+        className={ css('ms-FocusZone', rootProps && rootProps.className ) }
```
Comment (dzearing) [inline src/components/FocusZone/FocusZone.tsx]: I think this is wrong.

I think this should be className from the props, not from rootProps.

Comment (cliffkoh) [inline src/components/FocusZone/FocusZone.tsx]: Thank you :)

File: src/components/Button/Button.tsx [hunk 8]
```diff
@@ -30,18 +29,19 @@ export class Button extends React.Component<IButtonProps, IButtonState> {
...
 
-    const className = css(this.props.className, 'ms-Button', {
+    const className = css((this.props.className), 'ms-Button', {
       'ms-Button--primary': buttonType === ButtonType.primary,
       'ms-Button--hero': buttonType === ButtonType.hero,
       'ms-Button--compound': buttonType === ButtonType.compound,
       'ms-Button--command': buttonType === ButtonType.command,
       'ms-Button--icon': buttonType === ButtonType.icon,
-      'disabled': (elementType === ElementType.anchor && disabled)
+      'disabled': (!renderAsAnchor && disabled)
```
Comment (dzearing) [inline src/components/Button/Button.tsx]: not sure why you have the ! in there. that doesn't seem like the same logic as before.

File: src/components/Button/Button.tsx [hunk 9]
```diff
@@ -30,18 +29,19 @@ export class Button extends React.Component<IButtonProps, IButtonState> {
   }
 
   public render(): JSX.Element {
-    let { buttonType, children, icon, description, ariaLabel, ariaDescription, elementType, disabled } = this.props;
+    let { buttonType, children, icon, description, ariaLabel, ariaDescription, href, disabled } = this.props;
     let { labelId, descriptionId, ariaDescriptionId } = this.state;
 
-    const tag = this.props.elementType === ElementType.button ? 'button' : 'a';
+    const renderAsAnchor: boolean = !!href;
+    const tag = !renderAsAnchor ? 'button' : 'a';
```
Comment (dzearing) [inline src/components/Button/Button.tsx]: why not just remove the ! and flip the conditions. :)

File: src/components/Button/Button.tsx (hunk 8 above)
Comment (cliffkoh) [inline src/components/Button/Button.tsx]: I didn't even realize it because I wrote this reasoning about what it should be and not what was there. What was there appears to be buggy in the first place 😦

**Comment by dzearing** [comment]
Approved

**Comment by kmees** [comment]
@cliffkoh @dzearing do you really want to stick to the `rootProps` approach? Spreading the extraneous props is a much better approach for people using the components in my opinion. Again `inputProps` is fine, but `rootProps` doesn't really add anything.

For comparison, other button implementations that spread the props:
- [x] [react-toolbox](https://github.com/react-toolbox/react-toolbox/blob/dev/components/button/Button.js) 
- [x] [react-mdl](https://github.com/tleunen/react-mdl/blob/master/src/Button.js#L25)
- [x] [react-bootstrap](https://github.com/react-bootstrap/react-bootstrap/blob/master/src/Button.js#L56-L57)
- [x] [react-wigets](https://github.com/jquense/react-widgets/blob/master/src/Button.jsx#L7-L16)

**Comment by dzearing** [comment]
The core issue is typescript's lack of the "everything else" in destructuring that's causing some difficulty. But there also exists the same problem in other components, where root isn't EXACTLY what we should be targeting.

Because of this, solutions to polyfill this behavior (like a "copy all values except these keys" helper) would be really fragile. As soon as someone adds a prop without updating the key exemption list would cause more React warnings. It is difficult to predict that you need to keep the exemption keys updated.

Another issue is consistency. If we use inputProps, but we don't follow that convention with root, it feels less consistent. It would be nice to explicitly namespace the buckets of stuff we plan to mix in.

Another issue, a little less interesting but worth mentioning, is documentation. It is much easier to document "rootProps, which is of type React.HTMLProps<HTMLDivElement>, gets mixed into the root div", than "any props provided in this interface that we don't explicitly handle get mixed into the root div".

So the compromise here is: for props that ergonomically make sense to have in the interface, let's explicitly put them in the interface, document them, and mix them into the root. For the ones we don't have a common use case for, but someone will, let's give them an escape hatch. This solution scales beyond the root element; for more complex components, we can use this to target props to specific elements.

Let me know what your thoughts are. I know it feels a little off. Working around the typescript issue seems concerning to me.

**Comment by kmees** [comment]
Just because the implementation is a bit harder in typescript (for now) shouldn't be the main reason to change the public API of a component. Yeah it sucks to have that kind of workaround but overlooked properties 'only' generate warnings and shouldn't break anything. Also, those errors are easy to catch with unit tests.

I can understand the issue with documentation and typing. I haven't really used TS so i don't know if having a typed `rootProps` is useful or not. But i know that eslint doesn't work on something like `rootProps` (minor problem i guess). Prefacing every property documentation with "Everything not listed here is passed to the underlying component / root element" might be enough?

For the complex components, do you want to introduce more `xxxProps` and `yyyProps` for subcomponent xxx and yyy or do you want to pass parts of `rootProps` to different subcomponents? I don't understand exactly what you mean. In either case i dont see the benefit of having `rootProps` additionally to `xxxProps`, ... or splitting the 'rest' of the remaining props.

TL;DR the typescript issue shouldn't be the main reason to introduce `rootProps` :)

**Comment by dzearing** [comment]
Sorry, not complaining that it's harder, we can deal with that, but not sure it's the right thing.

I agree we can probably deal with the fragility; you're right, not much bad stuff can happen and unit tests can help catch mistakes.

There are other consistency reasons why we'd want to do this.

First, let's assume this isn't the common case. The common cases should be covered by top level props so that controls feel ergonomic for the most part. We're talking about the cases we may not normally handle.

Second, let's ignore button. Button is an easy passthrough. Let's focus on Checkbox, which currently has this composite DOM shape:

```jsx
<div className='ms-Checkbox'>
  <input type='checkbox' />
  <label>foo</label>
</div>
```

Say that we don't handle onFocus. When the checkbox is focused, the consumer wants to do something in the app. But shoot, we forgot to add that to our props interface.

Intuitively they would do this:

```
<Checkbox onFocus={ ... } />
```

If we just blindly mixed in unrecognized props into the root, we would be doing the wrong thing here. The focus handler should actually be on the input element, but instead goes to the unfocusable DIV. Same goes for just about everything else.

However using something like inputProps instead, we would first get a compiler error using TypeScript that onFocus is not defined. They would find the inputProps in documentation and pass it in that way. 

Alternatively we could try to do the right thing for every possible prop (like, onFocus was really meant for the input, onMouseOver maybe was meant for the whole thing? title was meant for the input, but maybe className was for the root? This doesn't seem super intuitive and sounds like huge bundle sizes and lots of bugs.

Let's also take another example, just composite controls in general. A command bar hosts a SearchBox, a CommandList for left side commands, and a CommandList for right side commands. This contract approach scales.

```typescript
ICommandBarProps {
  searchBoxProps: ISearchBoxProps;
  nearCommandListProps: ICommandListProps;
  farCommandListProps: ICommandListProps;
}
```

renders:
```
<div className='ms-CommandBar' { ...rootProps }>
  <SearchBox { ...searchBoxProps } />
  <CommandList { ...nearCommandListProps } />
  <CommandList { ...farCommandListProps } />
</div>
```

So, is there really a special distinction we'd want to have for root props? Is it really what you want, all of the time? Will it cause more issues than help? Should we care if it's used only exceptional scenarios?

**Comment by kmees** [comment]
The issue with something like 
```jsx
<div className='ms-CommandBar' { ...rootProps }>
  <SearchBox { ...searchBoxProps } />
  <CommandList { ...nearCommandListProps } />
  <CommandList { ...farCommandListProps } />
</div>
```
is that it exposes a lot of internal structure which is usually not a good idea and i can't really think of any situations where you want to pass in those props (except if you want to 'butcher' the component which will probably break the design guidelines?). For example if this component can be disabled, the component should be responsible for disabling all subcomponents etc. Some frameworks expose props for only adding `className`s to subcomponents instead of generic props since that is the most common use.

Also for the Checkbox, one could also follow the philosophy of 'passing down all additional props to the element where it makes most sense'. Thats basically what I did in my implementation.

```jsx
const ChoiceField = ({
  className, disabled, id, label, name, required, type, checked, ...props
}) => {
  const inputId = `ChoiceField_${id || name || Date.now()}_input`

  return (
    <div data-fabric="ChoiceField"
      id={id}
      className={className}
      styleName="ms-ChoiceField">
      <input {...props}
        styleName="ms-ChoiceField-input"
        children={null}
        id={inputId}
        name={name}
        type={type}
        disabled={disabled}
        checked={isDefined(checked) ? checked : false} />
      <label styleName="ms-ChoiceField-field"
        htmlFor={inputId}>
        <Label componentClass="span" styleName="ms-Label" required={required}>{label}</Label>
      </label>
    </div>
  )
}
```
You could add a `rootProps` to only those components (components that don't pass additional props to the root) as an escape hatch. The most common use case is probably `className` again which is way I explicitly added that to the root.

Another thing that came to mind about `rootProps` is that how do you handle duplicates, i.e props that are defined on the component itself but also valid on `rootProps` (due to being defined in React.HTMLProps)?

Probably something like this?
```jsx
import cx from 'classnames'
const Component = ({ className, rootProps }) => (
  <div className={cx('my-class', className, rootProps.className)} />
)
```
But this has issues with properties that are not composable
```jsx
const Component = ({ id, rootProps }) => (
  <div id={rootProps.id || id} />
) 
```
This is probably a very constructed example but just to give some food for thought.

Really hard to get this consistent i guess :)
